matscipy/dislocation.py

```diff
@@ -2150,6 +2152,242 @@ def check_duplicates(atoms, distance=0.1):
+        n = np.einsum('ij,j', A, n)
+        xi = np.einsum('ij,j', A, xi)
+        b = np.einsum('ij,j', A, burgers)
+        
+        # define elastic constant in Voigt notation
+        C_v = cubic_to_Voigt_6x6(C11, C12, C44)
+        # convert Voigt to tensor
+        C = Voigt_6x6_to_full_3x3x3x3(C_v)
+        # rotate elastic matrix
+        cijkl = np.einsum('ig,jh,ghmn,km,ln', \
+            A, A, C, A, A)
+    
+        # solve the Stroh sextic formalism: the same as Stroh.py from atomman
+        # this part can also be replaced by calling Stroh.py from atomman
+        mm = np.einsum('i,ijkl,l', m, cijkl, m)
+        mn = np.einsum('i,ijkl,l', m, cijkl, n)
+        nm = np.einsum('i,ijkl,l', n, cijkl, m)
+        nn = np.einsum('i,ijkl,l', n, cijkl, n)
+
+        # TODO: Replace with np.linalg.solve as it is more stable than np.linalg.inv
+        nninv = np.linalg.inv(nn)
+        mn_nninv = np.dot(mn, nninv)      
+        #mn_nninv_2 = np.linalg.solve(nn.T, mn.T).T # works, consistent with mn_nninv
+        #nninv_2 = np.linalg.solve(mn, mn_nninv_2) # TODO: Resolve LinAlgError: Singular matrix
+
+        N = np.zeros((6,6), dtype=float)
+        N[0:3,0:3] = -np.dot(nninv, nm)
+        N[0:3,3:6] = -nninv
+        N[3:6,0:3] = -(np.dot(mn_nninv, nm) - mm)
+        N[3:6,3:6] = -mn_nninv
+
+        # slove the eigenvalue problem
+        Np, Nv = np.linalg.eig(N)
+
+        # The eigenvector Nv contains the vectors A and L.
+        # Normalize A and L, such that 2*A*L=1 
+        for i in range(0,6):
+            norm = 2.0 * np.dot(Nv[0:3, i], Nv[3:6, i])
+            Nv[0:3, i] /= np.sqrt(norm)
+            Nv[3:6, i] /= np.sqrt(norm)
+
+        # Store key variables
+        self.m = m
+        self.n = n
+        self.xi = xi
+        self.b = b
+        self.Np = Np
+        self.Nv = Nv
+
+
+    def stroh_solve(self, coordinates): 
+        """
+        Displacement field of a straight dislocation. Currently only for 2D, can be extended.
+        Parameters
+        ----------
+        coordindates : array
+            Cartesian atomic coordinates with respect to the dislocation core.
+        Returns
+        -------
+        disp : array
+            Stroh displacements along the crack running direction (axes[0]) and crack plane normal (axes[1]).
+        """
+        
+        # calculation
+        signs = np.sign(np.imag(self.Np))
+        signs[np.where(signs==0.0)] = 1.0
+        A = self.Nv[0:3,:]
+        L = self.Nv[3:6,:]
+        D = np.einsum('i,ij', self.b, L)
+        constant_factor = signs * A * D
+
+        eta = (np.expand_dims(np.einsum('i,ji', self.m, coordinates), axis=1)
+            + np.outer(np.einsum('i,ji', self.n, coordinates), self.Np))
+
+        # get the displacements
+        disp = ((1.0/(2.0 * np.pi * 1.0j))
+            * np.einsum('ij,kj', np.log(eta), constant_factor))
+
+        return np.real(disp)
+    
+
+    def displacements(self, bulk, center=None, self_consistent=False,
```

Comment:
Is this identical to the SCF loop in `CubicCrystalDislocation`? If so could we avoid repeated code by putting it into a common base class?

Reply:
Yes it's the same SCF loop. I've now integrated the AD class with CCD, so that AD can use the self consistent solver via CCD.